java-frontend/src/main/java/org/sonar/java/cfg/CFGLoop.java

```diff
@@ -0,0 +1,178 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class CFGLoop {
+
+  private final CFG.Block startingBlock;
+  private final Set<CFG.Block> blocks = new HashSet<>();
+  private final Set<CFG.Block> successors = new HashSet<>();
+
+  private CFGLoop(CFG.Block block, Map<Tree, CFGLoop> container) {
+    container.put(block.terminator(), this);
+    startingBlock = block;
+    Block loopFirstBlock = block.trueBlock();
+    if (loopFirstBlock == null) {
+      // Special case where no condition is given in FOR loop: only one successor specified
+      loopFirstBlock = block.successors().iterator().next();
+    }
+    collectBlocks(loopFirstBlock, new HashSet<CFG.Block>(), container);
+    successors.addAll(block.successors());
+    successors.remove(block.falseBlock());
+    collectWaysOut(container);
+  }
+
+  public CFG.Block startingBlock() {
+    return startingBlock;
+  }
+
+  public Collection<CFG.Block> blocks() {
+    return sorted(blocks);
+  }
+
+  public Collection<CFG.Block> successors() {
+    return sorted(successors);
+  }
+
+  public boolean hasNoWayOut() {
+    return successors.isEmpty();
+  }
+
+  private boolean collectBlocks(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
+    if (blocks.contains(block)) {
+      return true;
+    }
+    if (block == null || block == startingBlock || !visitedBlocks.add(block)) {
+      return false;
+    }
+    boolean answer = returnsToStart(block, visitedBlocks, container);
+    if (answer || isBreak(block)) {
+      blocks.add(block);
+    }
+    return answer;
+  }
+
+  private boolean returnsToStart(@Nonnull CFG.Block block, Set<CFG.Block> visitedBlocks, Map<Tree, CFGLoop> container) {
+    Set<Block> localSuccessors = localSuccessors(block, container);
+    if (localSuccessors == null) {
+      return true;
+    }
+    boolean answer = false;
+    for (CFG.Block successor : localSuccessors) {
+      if (startingBlock == successor) {
+        answer = true;
+      } else {
+        answer |= collectBlocks(successor, visitedBlocks, container);
+      }
+    }
+    return answer;
+  }
+
+  @CheckForNull
+  private Set<Block> localSuccessors(@Nonnull CFG.Block block, Map<Tree, CFGLoop> container) {
+    if (isStarting(block)) {
+      CFGLoop loop = container.get(block.terminator());
+      if (loop == null) {
+        loop = new CFGLoop(block, container);
+      }
+      Set<Block> loopSuccessors = new HashSet<>(loop.successors);
+      if (block.falseBlock() != null) {
+        loopSuccessors.add(block.falseBlock());
+      } else if (block.trueBlock() == null) {
+        // Special case where no condition is given in FOR loop: only one successor specified
+        return null;
+      }
+      return loopSuccessors;
+    }
+    return block.successors();
+  }
+
+  private boolean isBreak(@Nonnull CFG.Block block) {
+    return block.terminator() != null && block.terminator().is(Tree.Kind.BREAK_STATEMENT);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'terminator' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)